From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: "Jan Alexander Steffens (heftig)" <heftig@archlinux.org>
Date: Thu, 15 Oct 2026 23:06:46 +0200
Subject: [PATCH] meson: support installing agent-helper as socket-activated

Arch Linux wants the helper to be socket-activated statically.
---
 data/meson.build                              | 23 +++++++++++++++--
 ...r.socket => polkit-agent-helper.socket.in} |  4 +--
 meson.build                                   |  8 ++++++
 meson_options.txt                             |  1 +
 meson_post_install.py                         | 25 ++++++++-----------
 5 files changed, 43 insertions(+), 18 deletions(-)
 rename data/{polkit-agent-helper.socket => polkit-agent-helper.socket.in} (71%)

diff --git a/data/meson.build b/data/meson.build
index 9e35b2f97a69..a87815bfe40e 100644
--- a/data/meson.build
+++ b/data/meson.build
@@ -1,8 +1,16 @@
+if activated_helper == 'true'
+  # Comment out polkit-agent-helper.socket's Install section
+  helper_install = '#'
+else
+  helper_install = ''
+endif
+
 service_conf = {
   'libprivdir': pk_prefix / pk_libprivdir,
   'polkitd_user': polkitd_user,
   'polkitd_uid': polkitd_uid,
   'polkitd_gid': polkitd_gid,
+  'helper_install': helper_install,
 }
 
 configure_file(
@@ -56,11 +64,22 @@ if not get_option('libs-only')
     install_dir: systemdsystemunitdir,
   )
 
-  install_data(
-    'polkit-agent-helper.socket',
+  configure_file(
+    input: 'polkit-agent-helper.socket.in',
+    output: '@BASENAME@',
+    configuration: service_conf,
+    install: true,
     install_dir: systemdsystemunitdir,
   )
 
+  if activated_helper == 'true'
+    install_symlink(
+      'polkit-agent-helper.socket',
+      install_dir: systemdsystemunitdir / 'sockets.target.wants',
+      pointing_to: '../polkit-agent-helper.socket',
+    )
+  endif
+
   if polkitd_gid != '-'
     sysusers_template = 'polkit.conf.in-with-gid'
   else
diff --git a/data/polkit-agent-helper.socket b/data/polkit-agent-helper.socket.in
similarity index 71%
rename from data/polkit-agent-helper.socket
rename to data/polkit-agent-helper.socket.in
index e8fa80b5d952..c442e462ab4d 100644
--- a/data/polkit-agent-helper.socket
+++ b/data/polkit-agent-helper.socket.in
@@ -7,5 +7,5 @@ Accept=yes
 RemoveOnStop=yes
 ListenStream=/run/polkit/agent-helper.socket
 
-[Install]
-WantedBy=sockets.target
+@helper_install@[Install]
+@helper_install@WantedBy=sockets.target
diff --git a/meson.build b/meson.build
index 86a4196937ec..d5d69f4c0d0f 100644
--- a/meson.build
+++ b/meson.build
@@ -338,6 +338,12 @@ if privileged_group == ''
   endif
 endif
 
+# The way the helper is activated differs between distributions, set it accordingly if not specified
+activated_helper = get_option('activated_helper')
+if activated_helper == ''
+  activated_helper = 'false'
+endif
+
 pam_include = get_option('pam_include')
 if pam_include == ''
   if ['suse', 'solaris', 'openembedded'].contains(os_type)
@@ -420,6 +426,7 @@ if not libs_only
     pk_pkgsysconfdir,
     polkitd_user,
     polkitd_gid,
+    activated_helper,
   )
 endif
 
@@ -451,6 +458,7 @@ endif
 if polkitd_gid != '-'
   output += '        polkitd GID:              ' + polkitd_gid + '\n'
 endif
+output += '        Socket-activated helper:  ' + activated_helper + '\n'
 output += '        PAM support:              ' + enable_pam.to_string() + '\n\n'
 if libs_only
   output += '    !!! Only building polkit libraries, not polkitd !!!\n\n'
diff --git a/meson_options.txt b/meson_options.txt
index 4295a97938c5..8f9e48207787 100644
--- a/meson_options.txt
+++ b/meson_options.txt
@@ -6,6 +6,7 @@ option('polkitd_user', type: 'string', value: 'polkitd', description: 'Name of u
 option('polkitd_uid', type: 'string', value: '', description: 'UID for polkitd_user (default: - to allocate)')
 option('polkitd_gid', type: 'string', value: '', description: 'GID for polkitd_user\'s primary group (default: - to allocate)')
 option('privileged_group', type: 'string', value: '', description: 'Group to use for default privileged access (default: wheel)')
+option('activated_helper', type: 'combo', choices: ['false', 'true', ''], value: '', description: 'Enable socket-activated agent helper (default: false)')
 
 option('authfw', type: 'combo', choices: ['pam', 'shadow', 'bsdauth'], value: 'pam', description: 'Authentication framework (pam/shadow)')
 option('os_type', type: 'combo', choices: ['redhat', 'suse', 'gentoo', 'pardus', 'solaris', 'netbsd', 'lfs', 'openembedded', 'debian', ''], value: '', description: 'distribution or OS')
diff --git a/meson_post_install.py b/meson_post_install.py
index 00592a22210e..545c0cf64cae 100644
--- a/meson_post_install.py
+++ b/meson_post_install.py
@@ -21,44 +21,41 @@ pkglibdir = destdir_path(sys.argv[2])
 pkgsysconfdir = destdir_path(sys.argv[3])
 polkitd_user = sys.argv[4]
 polkitd_gid = sys.argv[5]
+activated_helper = sys.argv[6]
 
 if polkitd_gid == '-':
     try:
         polkitd_gid = pwd.getpwnam(polkitd_user).pw_gid
     except KeyError:
         polkitd_gid = None
 else:
     polkitd_gid = int(polkitd_gid)
 
 dst = os.path.join(bindir, 'pkexec')
 
 if os.geteuid() == 0:
     os.chown(dst, 0, -1)
     os.chmod(dst, 0o4755)
 else:
     print(f'Owner and mode of {dst} need to be setuid root (04755) after installation')
 
 dst = os.path.join(pkgsysconfdir, 'rules.d')
 
 if not os.path.exists(dst):
     os.makedirs(dst, mode=0o750)
     if os.geteuid() == 0 and polkitd_gid is not None:
         os.chown(dst, 0, polkitd_gid)
     else:
         print(f'Owner of {dst} needs to be set to root and group to {polkitd_user} after installation')
 
-# polkit-agent-helper-1 need to be setuid root because it's used to
-# authenticate not only the invoking user, but possibly also root
-# and/or other users.
-dst = os.path.join(pkglibdir, 'polkit-agent-helper-1')
+if activated_helper != 'true':
+    # polkit-agent-helper-1 need to be setuid root because it's used to
+    # authenticate not only the invoking user, but possibly also root
+    # and/or other users.
+    dst = os.path.join(pkglibdir, 'polkit-agent-helper-1')
 
-if os.geteuid() == 0:
-    os.chown(dst, 0, -1)
-    os.chmod(dst, 0o4755)
-else:
-    print(
-        'Owner and mode of {} need to be setuid root (04755) after '
-        'installation'.format(
-            dst,
-        )
-    )
+    if os.geteuid() == 0:
+        os.chown(dst, 0, -1)
+        os.chmod(dst, 0o4755)
+    else:
+        print(f'Owner and mode of {dst} need to be setuid root (04755) after installation')
